src/core/net/ip6.cpp

```diff
@@ -1000,14 +1000,12 @@ Error Ip6::HandlePayload(Header &           aIp6Header,
 Error Ip6::ProcessReceiveCallback(Message &          aMessage,
                                   const MessageInfo &aMessageInfo,
                                   uint8_t            aIpProto,
-                                  bool               aFromNcpHost,
                                   bool               aAllowReceiveFilter,
                                   Message::Ownership aMessageOwnership)
 {
     Error    error   = kErrorNone;
     Message *message = &aMessage;
 
-    VerifyOrExit(!aFromNcpHost, error = kErrorNoRoute);
```

Comment:
Thanks @simonlingoogle .

I understand that this is aimed at fixing the https://github.com/openthread/openthread/issues/7175, but I am wondering if this change may break some existing functionality/behaviors.

I think this `VerifyOrExit(!aFraomNcpHost)` here ensures that we never pass back the IPv6 message we get from host  back to it. I don't remember the details, but I recall we needed this line here to address a bug/issue early on. Without this I think we may end up in a loop between host netif and OT, passing the same message back and forth between them.

Reply:
I did some digging on this. I found 
- https://github.com/openthread/openthread/pull/1583 

This is the PR where this line was added (our naming and code style were somewhat different back then... 😄)

Reply:
Thanks for informing this ancient PR. :smile: 

I think the main issue is a potential message loop, but I think it should be fine:
- We are going to have #7107, help us address this issues:
  - Only message with OMR destinations and DUA destinations will be passed via TUN to Thread stack
    - For OMR destinations, the messages will should never be passed back to TUN (we can guarantee that with egress safeguard (which I believe is explicitly required by Thread Spec)).
    - For DUA destinations, the messages may be passed back to TUN if the DUA ND Proxying module has a bug or if another BR has outdated ARP caches. Even if it happens, the message will be passed via TUN to Thread, and via TUN back to kernel just once. Then kernel will forward the message to the another BR. There won't be a message loop in in such case. 
- TUN interface decreases TTL by 1 when forwarding IP6 messages, so TTL will make sure the message is eventually dropped even when loop really happens

So, there seems not going to be serious message loops. 

Thoughts?

Reply:
I think this  situation may not just be applicable to the BR devices (and/or NCP model or posix/linux TUN netif). I think it can happen on all device types (end-devices) and in other situations (e.g., with lwip netif, or other use-cases).

- Basically in the current code we ensure that any time user of OT calls `otIp6Send()` to send an IPv6 message, the OT stack will not pass it back to the user in a `otIp6ReceiveCallback` 
   - The `VerifyOrExit(!aFromNcpHost)` ensures this. 
- I think with the change in this PR, we will  change this behavior. We should be careful about the impact of such a change.

I have not been following #7175 in detail and I am not sure if I understand how this change is fixing it (@simonlingoogle can you please help add a short explanation?). Does the #7175 require that "message given from `otIp6Send()` to be passed in `otIp6ReceiveCallback` callback"? If not, can we possibly fix this in a different way?

Reply:
> I think this situation may not just be applicable to the BR devices (and/or NCP model or posix/linux TUN netif). I think it can happen on all device types (end-devices) and in other situations (e.g., with lwip netif, or other use-cases).

Essentially, I think it's correct to forward the message equally no matter it's initiated in Thread stack or from Host. (@bukepo may have comments on it) So, this PR is a step in that direction. 

> Does the #7175 require that "message given from `otIp6Send()` to be passed in `otIp6ReceiveCallback` callback"?

Yes. In #7175, SBBR has subscribed an multicast address MA1, and the Host on Backbone Link is pining MA1. 
When SBBR receives the ping request to MA1 in **kernel**, kernel sends ping reply via wpan0 to Thread stack. Now, Thread stack will choose SBBR itself as the Border Router because the destination of ping reply is an on-link address, which SBBR has added as an external route. Currently, SBBR failed to foward the ping reply because:
1) SBBR can't forward the ping reply back to TUN due to `aFromNcpHost` checks
2) SBBR can't forward the ping reply to another BR because it naturally choose itself as the BR

So, SBBR need to either forward the ping reply back to TUN, or to another BR. 
This PR tries to fix 1) and I think the change is obvious and clean. 
I think @edmont tried to fix 2) in #7193 but it seemed to be quite complex. 

Note that 1) is the best route, while 2) is routing the message with much higher cost. So, 1) should be the correct behavior. 
So, I think this PR is indeed a bugfix.

Reply:
> Essentially, I think it's correct to forward the message equally no matter it's initiated in Thread stack or from Host. (@bukepo may have comments on it) So, this PR is a step in that direction.

Even if we argue that it may be correct to do so (and that it is up to next layer to filter messages), I think it will be a change to existing behavior and it may have impact (may have some unintended consequences). I suggest that we try to be careful about this.

I tried digging into older internal bugs/Jiras on what issues we saw related to this (if/when we allowed msgs from host to br passed back to the host):
- One use-case was to use thread netif (wpan0) as a low priority default route.
   - The use-case was, for example if WiFi is connected on device, we use that for all traffic (e.g. traffic to a service/cloud), but if WiFi is down, then we allow Thread netif to be used (in hope that the traffic may be able to routed in Thread mesh to a BR which can then forward it to the intended destination). 
   - That's where the loop issue could happen.
   - We can argue that such a use may not be valid/allowed (with new BR model) and/or we need additional filtering (e.g. ingress filtering) to allow this. But would be good to consider it.
- I recall there were some issues related to ICMP redirect msgs (host/tun netif sending them down to OT and when passed back causes some issue on host side itself) - I think we added filtering in `wpantund` for this.
- I recall we saw similar (looping of msg) issues on RTOS type of device which used lwip for IP stack.

-------

Thanks for explaining the MA1 problem. 

Does the OT stack know MA1 address in this case. Can we consider adding mechansim to allow passing a message back to host only if the address is the known MA1? Would this work and help address #7175?

Reply:
Thanks for providing the background. 

Agree that disallowing forwarding message with `aFromNcpHost` to Host can avoid potential message loops. 
However, I think recent Border Router development allows us to be more comfortable allowing that:
- Security safeguard only allows a strict set of destinations (OMR & DUA) to be enter Thread, creating a much safer environment.
- With https://github.com/openthread/ot-br-posix/pull/1063, we actually don't allow forwarding from `wpan0` to `wpan0` (@superwhd correct me if I am wrong). So, any message that Thread stack forwards to `wpan0`, kernel will never be forwarding back to `wpan0`. This effectively eliminates any message loops. For any other platforms, they should be implementing similar safeguard to be inline with Thread Spec. 

Another view point is that `otIp6Send` is sending a message via Thread stack. So, the BR functionality of a device should be able to forward the message to Host no matter the message is initiated by itself or another Thread device. Current `aFromNcpHost` checking actually breaks this functionality, making the behavior of `otIp6Send` inconsistent on BR and non-BR devices.

Reply:
I feel like we can **disallow forwarding from wpan0 to wpan0** to eliminate message loops.

https://github.com/openthread/ot-br-posix/pull/1063 only allows messages with these destinations to be forwarded to Thread:
- OMR: messages with OMR dst should not be forwarded to Host in the first place, so fine to block them. 
- DUA: messages with DUA dst should not be forwarded back to wpan0 if they are forwarded to Host, so fine to block them.
- Multicast: messages with Multicast dst should not be forwarded back to wpan0 if they are forwarded to Host, so fine to block them. 

Thoughts? @superwhd @librasungirl

Reply:
SGTM to disallow that forwarding.

Reply:
@abtink 

Agree that it's important to avoid potential message loops. 

Now that we can disallow forwarding from wpan0 to wpan0 by firewall rules, it seems more safe to allow forwarding to Host when the message is from Host. 

One problem with messages from Host is that we can't tell if the message is 1) forwarded by Host from infrastructure link, or 2) if it's initiated by any other Host program locally. Generally, we need to disallow forwarding to Host for case 1) but allow forwarding for case 2). But since we are unable to tell where the messages are from, allowing forwarding to Host should be more conservative. 

Thoughts?

Reply:
I understand the concerns about message loops but I agree with @simonlingoogle that those are greatly mitigated.

What is certain is that we need a solution for the case described in https://github.com/openthread/openthread/issues/7175, which is blocking Thread 1.2 certification program, and we don't have any better alternative at the moment.

Reply:
Here are my thoughts and suggestions to move this forward:

- This PR as is is changing a key existing behavior, i.e., OT does not pass back the message it receives from user `otIp6Send ` back to it in the callback `otIp6ReceiveCallback`). 
   - Our experiences from past (also my gut-feeling) point that such a change may have unintended consequences (looping is obvious thing we see but there may be other issues). 

- My first suggestion is to consider if a more surgical fix can be added for #7175.

- If we want to move forward with the approach in this PR here is my suggestion:
  - The current PR fully removes the behavior from the OT code.
  - I would suggest adding this change under a `CONFIG` option or may be even adding an API to determine the beavior (i.e. a config or API that either allows or disallows passing back user's message back to it in the callback). This way, the risk is lowered.

Reply:
@abtink 

I have restructured this PR so that it does not remove `aFromNcpHost` logic at all. However, we change `otIp6Send` to use `aFromNcpHost = false` instead. 

Please take a look.

Reply:
Thanks. I think this can work.